ksqldb-engine/src/main/java/io/confluent/ksql/util/QueryMetadataImpl.java

```diff
@@ -447,6 +447,20 @@ public void resume() {
     listener.onResume(this);
   }
 
+  private QueryError.Type recursiveClassification(final Throwable throwable, final int depthLeft) {
+    // depthLeft is used just here to safeguard against infinite recursion in the case of
+    // self-referencing exceptions
+    if (throwable != null && depthLeft > 0) {
+      final QueryError.Type errorType = errorClassifier.classify(throwable);
+      if (errorType != QueryError.Type.UNKNOWN) {
+        return errorType;
+      } else {
+        return recursiveClassification(throwable.getCause(), depthLeft - 1);
+      }
+    }
+    return QueryError.Type.UNKNOWN;
```

Comment:
Would `Throwables.getCausalChain` help instead of doing recursion and using a depthLeft?

```
for (Throwable t : Throwables.getCausalChain(throwable)) {
    final QueryError.Type errorType = errorClassifier.classify(t);
    if (errorType != QueryError.Type.UNKNOWN) {
      return errorType;
    }
}

return QueryError.Type.UNKNOWN;
```

Reply:
Done